### PR TITLE
Improve docs and tests

### PR DESCRIPTION
**Pull request by wbloszyk**
## Subject

Improve docs and tests.

I am targeting this branch, because this change respect BC.

### CODE REVIEW

File: docs/reference/advanced_configuration.rst [hunk 1]
```diff
@@ -33,3 +33,8 @@ Advanced Configuration
                     class: Sonata\ClassificationBundle\Admin\ContextAdmin
                     controller: SonataAdminBundle:CRUD
                     translation: SonataClassificationBundle
+
+        jms_serializer:
+            metadata:
+                directories:
+                - { name: 'sonata_classification', path: "%kernel.project_dir%/vendor/sonata-project/classification-bundle/src/Resources/config/serializer", namespace_prefix: 'Sonata\ClassificationBundle' }
```
Comment (VincentLanglet) [inline docs/reference/advanced_configuration.rst:40]:
```suggestion
                    - { name: 'sonata_classification', path: "%kernel.project_dir%/vendor/sonata-project/classification-bundle/src/Resources/config/serializer", namespace_prefix: 'Sonata\ClassificationBundle' }
```

File: src/Controller/Api/CategoryController.php [hunk 2]
```diff
@@ -275,7 +275,7 @@ protected function getCategory($id)
         $category = $this->categoryManager->find($id);
 
         if (null === $category) {
-            throw new NotFoundHttpException(sprintf('Category (%s) not found', $id));
+            throw new NotFoundHttpException(sprintf('Category not found for identifier %s.', var_export($id, true)));
```
Comment (OskarStark) [inline src/Controller/Api/CategoryController.php:278]: Shall we use a named ctor?

```php
NotFoundHttpException implements NotFoundHttpException
{
    public static function forIdentifier(string $identifier)
....
}
```

??

Comment (wbloszyk) [reply src/Controller/Api/CategoryController.php:278]: This Expection should be provide in 7 bundles (https://github.com/sonata-project/dev-kit/issues/778). IMO it is another reason to create `sonata-project/contracts`. 

For now I will keep current version.

File: docs/reference/advanced_configuration.rst [hunk 3]
```diff
@@ -33,3 +33,9 @@ Advanced Configuration
                     class: Sonata\ClassificationBundle\Admin\ContextAdmin
                     controller: SonataAdminBundle:CRUD
                     translation: SonataClassificationBundle
+
+        jms_serializer:
+            metadata:
+                directories:
+                    - { name: 'sonata_datagrid', path: "%kernel.project_dir%/vendor/sonata-project/datagrid-bundle/src/Resources/config/serializer", namespace_prefix: 'Sonata\DatagridBundle' }
+                    - { name: 'sonata_classification', path: "%kernel.project_dir%/vendor/sonata-project/classification-bundle/src/Resources/config/serializer", namespace_prefix: 'Sonata\ClassificationBundle' }
```
Comment (greg0ire) [inline docs/reference/advanced_configuration.rst:41]:
```suggestion
                    -
                        name: 'sonata_datagrid'
                        path: "%kernel.project_dir%/vendor/sonata-project/datagrid-bundle/src/Resources/config/serializer"
                        namespace_prefix: 'Sonata\DatagridBundle'
                    -
                        name: 'sonata_classification'
                        path: "%kernel.project_dir%/vendor/sonata-project/classification-bundle/src/Resources/config/serializer"
                        namespace_prefix: 'Sonata\ClassificationBundle'
```

File: docs/reference/api.rst [hunk 4]
```diff
@@ -26,11 +26,15 @@ Here's the configuration we used, you may adapt it to your needs:
...
         request: { converters: true }
-        format_listener:
-            rules:
-                - { path: '^/', priorities: ['json'], fallback_format: json, prefer_extension: false }
+
+    jms_serializer:
+        metadata:
+            directories:
+                - { name: 'sonata_datagrid', path: "%kernel.project_dir%/vendor/sonata-project/datagrid-bundle/src/Resources/config/serializer", namespace_prefix: 'Sonata\DatagridBundle' }
+                - { name: 'sonata_classification', path: "%kernel.project_dir%/vendor/sonata-project/classification-bundle/src/Resources/config/serializer", namespace_prefix: 'Sonata\ClassificationBundle' }
```
Comment (greg0ire) [inline docs/reference/api.rst:37]:
```suggestion
                -
                    name: 'sonata_datagrid'
                    path: "%kernel.project_dir%/vendor/sonata-project/datagrid-bundle/src/Resources/config/serializer"
                    namespace_prefix: 'Sonata\DatagridBundle'
                -
                     name: 'sonata_classification'
                     path: "%kernel.project_dir%/vendor/sonata-project/classification-bundle/src/Resources/config/serializer"
                     namespace_prefix: 'Sonata\ClassificationBundle'
```

File: tests/Controller/Api/CategoryControllerTest.php [hunk 5]
```diff
@@ -54,12 +54,28 @@ public function testGetCategoryAction(): void
...
 
-        $this->createCategoryController()->getCategoryAction(42);
+    /**
+     * @phpstan-return list<array{mixed, string}>
+     */
+    public function getIdsForNotFound(): array
+    {
+        return [
+            [42, 'Category not found for identifier 42.'],
+            ['42', 'Category not found for identifier \'42\'.'],
```
Comment (greg0ire) [inline tests/Controller/Api/CategoryControllerTest.php:75]:
```suggestion
            ['42', "Category not found for identifier '42'."],
```

File: tests/Controller/Api/CategoryControllerTest.php [hunk 6]
```diff
@@ -54,12 +54,28 @@ public function testGetCategoryAction(): void
...
+            [null, 'Category not found for identifier NULL.'],
+            ['', 'Category not found for identifier \'\'.'],
```
Comment (greg0ire) [inline tests/Controller/Api/CategoryControllerTest.php:77]:
```suggestion
            ['', "Category not found for identifier ''."],
```

File: tests/Controller/Api/CollectionControllerTest.php [hunk 7]
```diff
@@ -54,12 +54,28 @@ public function testGetCollectionAction(): void
...
 
-        $this->createCollectionController()->getCollectionAction(42);
+    /**
+     * @phpstan-return list<array{mixed, string}>
+     */
+    public function getIdsForNotFound(): array
+    {
+        return [
+            [42, 'Collection not found for identifier 42.'],
+            ['42', 'Collection not found for identifier \'42\'.'],
```
Comment (greg0ire) [inline tests/Controller/Api/CollectionControllerTest.php:75]:
```suggestion
            ['42', "Collection not found for identifier '42'."],
```

File: tests/Controller/Api/CollectionControllerTest.php [hunk 8]
```diff
@@ -54,12 +54,28 @@ public function testGetCollectionAction(): void
...
+            [null, 'Collection not found for identifier NULL.'],
+            ['', 'Collection not found for identifier \'\'.'],
```
Comment (greg0ire) [inline tests/Controller/Api/CollectionControllerTest.php:77]:
```suggestion
            ['', "Collection not found for identifier ''."],
```

File: tests/Controller/Api/ContextControllerTest.php [hunk 9]
```diff
@@ -54,14 +54,33 @@ public function testGetContextAction(): void
...
     }
 
+    /**
+     * @phpstan-return list<array{mixed, string}>
+     */
+    public function getIdsForNotFound(): array
+    {
+        return [
+            [42, 'Context not found for identifier 42.'],
+            ['42', 'Context not found for identifier \'42\'.'],
```
Comment (greg0ire) [inline tests/Controller/Api/ContextControllerTest.php:75]:
```suggestion
            ['42', "Context not found for identifier '42'."],
```

File: tests/Controller/Api/ContextControllerTest.php [hunk 10]
```diff
@@ -54,14 +54,33 @@ public function testGetContextAction(): void
...
+            [null, 'Context not found for identifier NULL.'],
+            ['', 'Context not found for identifier \'\'.'],
```
Comment (greg0ire) [inline tests/Controller/Api/ContextControllerTest.php:77]:
```suggestion
            ['', "Context not found for identifier ''."],
```

File: tests/Controller/Api/TagControllerTest.php [hunk 11]
```diff
@@ -54,12 +54,28 @@ public function testGetTagAction(): void
...
 
-        $this->createTagController()->getTagAction(42);
+    /**
+     * @phpstan-return list<array{mixed, string}>
+     */
+    public function getIdsForNotFound(): array
+    {
+        return [
+            [42, 'Tag not found for identifier 42.'],
+            ['42', 'Tag not found for identifier \'42\'.'],
```
Comment (greg0ire) [inline tests/Controller/Api/TagControllerTest.php:75]:
```suggestion
            ['42', "Tag not found for identifier '42'."],
```

File: tests/Controller/Api/TagControllerTest.php [hunk 12]
```diff
@@ -54,12 +54,28 @@ public function testGetTagAction(): void
...
+            [null, 'Tag not found for identifier NULL.'],
+            ['', 'Tag not found for identifier \'\'.'],
```
Comment (greg0ire) [inline tests/Controller/Api/TagControllerTest.php:77]:
```suggestion
            ['', "Tag not found for identifier ''."],
```

File: docs/reference/api.rst (hunk 4 above)
Comment (phansys) [reply docs/reference/api.rst:37]: There seems to be an indentation issue.

Comment (greg0ire) [reply docs/reference/api.rst:37]: Where? I don't see it…

Comment (phansys) [reply docs/reference/api.rst:37]:
```patch
+                -
+                    name: 'sonata_datagrid'
+                    path: "%kernel.project_dir%/vendor/sonata-project/datagrid-bundle/src/Resources/config/serializer"
+                    namespace_prefix: 'Sonata\DatagridBundle'
+                -
+                    ·name: 'sonata_classification'
+                    ·path: "%kernel.project_dir%/vendor/sonata-project/classification-bundle/src/Resources/config/serializer"
+                    ·namespace_prefix: 'Sonata\ClassificationBundle'
```

Comment (greg0ire) [reply docs/reference/api.rst:37]: Oh indeed! Thanks, fixed!

File: docs/reference/advanced_configuration.rst (hunk 3 above)
Comment (wbloszyk) [reply docs/reference/advanced_configuration.rst:41]: https://symfony.com/doc/3.x/bundles/NelmioApiDocBundle/alternative_names.html

Symfony use this notation, should we too?

Comment (wbloszyk) [reply docs/reference/advanced_configuration.rst]: Anyway this should be move to jms_serializer tutorial. Removed here.

**Pull request closed without merging**